testing/endtoend/params/params.go

```diff
@@ -108,28 +194,125 @@ func InitMultiClient(beaconNodeCount int, lighthouseNodeCount int) error {
+	if err != nil {
+		return err
+	}
+	prysmBeaconNodePprofPort, err := port(PrysmBeaconNodePprofPort, testTotalShards, testShardIndex, &existingRegistrations)
+	if err != nil {
+		return err
+	}
+	lighthouseBeaconNodeP2PPort, err := port(LighthouseBeaconNodeP2PPort, testTotalShards, testShardIndex, &existingRegistrations)
+	if err != nil {
+		return err
+	}
+	lighthouseBeaconNodeHTTPPort, err := port(LighthouseBeaconNodeHTTPPort, testTotalShards, testShardIndex, &existingRegistrations)
+	if err != nil {
+		return err
+	}
+	lighthouseBeaconNodeMetricsPort, err := port(LighthouseBeaconNodeMetricsPort, testTotalShards, testShardIndex, &existingRegistrations)
+	if err != nil {
+		return err
+	}
+	validatorGatewayPort, err := port(ValidatorGatewayPort, testTotalShards, testShardIndex, &existingRegistrations)
+	if err != nil {
+		return err
+	}
+	validatorMetricsPort, err := port(ValidatorMetricsPort, testTotalShards, testShardIndex, &existingRegistrations)
+	if err != nil {
+		return err
+	}
+	testPorts := &ports{
+		BootNodePort:                    bootnodePort,
+		BootNodeMetricsPort:             bootnodeMetricsPort,
+		Eth1RPCPort:                     eth1RPCPort,
+		Eth1WSPort:                      eth1WSPort,
+		PrysmBeaconNodeRPCPort:          prysmBeaconNodeRPCPort,
+		PrysmBeaconNodeUDPPort:          prysmBeaconNodeUDPPort,
+		PrysmBeaconNodeTCPPort:          prysmBeaconNodeTCPPort,
+		PrysmBeaconNodeGatewayPort:      prysmBeaconNodeGatewayPort,
+		PrysmBeaconNodeMetricsPort:      prysmBeaconNodeMetricsPort,
+		PrysmBeaconNodePprofPort:        prysmBeaconNodePprofPort,
+		LighthouseBeaconNodeP2PPort:     lighthouseBeaconNodeP2PPort,
+		LighthouseBeaconNodeHTTPPort:    lighthouseBeaconNodeHTTPPort,
+		LighthouseBeaconNodeMetricsPort: lighthouseBeaconNodeMetricsPort,
+		ValidatorMetricsPort:            validatorMetricsPort,
+		ValidatorGatewayPort:            validatorGatewayPort,
+	}
 
 	TestParams = &params{
-		TestPath:                  testPath,
+		TestPath:                  filepath.Join(testPath, fmt.Sprintf("shard-%d", testShardIndex)),
 		LogPath:                   logPath,
-		TestShardIndex:            testIndex,
+		TestShardIndex:            testShardIndex,
 		BeaconNodeCount:           beaconNodeCount,
 		LighthouseBeaconNodeCount: lighthouseNodeCount,
-		Eth1RPCPort:               3100 + testIndex*100, // Multiplying 100 here so the test index doesn't conflict with the other node ports.
-		BootNodePort:              4100 + testIndex*100,
-		BeaconNodeRPCPort:         4150 + testIndex*100,
-		BeaconNodeMetricsPort:     5100 + testIndex*100,
-		ValidatorMetricsPort:      6100 + testIndex*100,
-		ValidatorGatewayPort:      7150 + testIndex*100,
+		Ports:                     testPorts,
 	}
 	return nil
 }
+
+// port returns a safe port number based on the seed and shard data.
+func port(seed, shardCount, shardIndex int, existingRegistrations *[]int) (int, error) {
```

Comment:
a small unit test for this method would be good for general correctness

Reply:
Yeah, it already found a bug...